example/common/flat_write_stream.hpp

```diff
@@ -0,0 +1,554 @@
+                necessary, ownership of the underlying memory blocks is
+                retained by the caller, which must guarantee that they
+                remain valid until the handler is called.
+                @param handler The handler to be called when the operation
+                completes. Copies will be made of the handler as required.
+                The equivalent function signature of the handler must be:
+                @code void handler(
+                    error_code const& error,      // result of operation
+                    std::size_t bytes_transferred // number of bytes transferred
+                ); @endcode
+                Regardless of whether the asynchronous operation completes
+                immediately or not, the handler will not be invoked from within
+                this function. Invocation of the handler will be performed in a
+                manner equivalent to using `boost::asio::io_service::post`.
+            */
+            template<class ConstBufferSequence, class WriteHandler>
+#if BEAST_DOXYGEN
+            void_or_deduced
+#else
+            async_return_type<WriteHandler, void(error_code)>
+#endif
+            async_write_some(ConstBufferSequence const &buffers,
+                             WriteHandler &&handler)
+            {
+                static_assert(is_async_write_stream<next_layer_type>::value,
+                              "AsyncWriteStream requirements not met");
+                static_assert(is_const_buffer_sequence<
+                                      ConstBufferSequence>::value,
+                              "ConstBufferSequence requirements not met");
+                static_assert(is_completion_handler<WriteHandler,
+                                      void(error_code, std::size_t)>::value,
+                              "WriteHandler requirements not met");
+
+                auto size = boost::asio::buffer_size(buffers);
+
+                std::cout << "write_some buffer size: " << size << std::endl;   //todo: remove
+
+                boost::beast::async_completion<WriteHandler,
+                        void(boost::beast::error_code, size_t)>
+                        init{handler};
+
+                if(size > 0) {
+                    auto iter = find_single(buffers);
+                    if(iter != buffers.end()) {
+                        write_some_op<
+                                ConstBufferSequence,
+                                boost::beast::handler_type<
+                                        WriteHandler, void(boost::beast::error_code, std::size_t)>> op(handler, *this, boost::asio::const_buffers_1{*iter}, std::shared_ptr<char>(nullptr));
+
+                        op(boost::beast::error_code{}, 0);
+                        return init.result.get();
+                    }
+                    else {
+                        if(size <= max_stack_size) {
+                            write_some_op<
```

Comment:
You can't flatten on the stack in the asynchronous case, because the initiating function returns before the buffer is sent!